Out-of-core state backend for JDBC databases

gyfora opened the pull request:
## Out-of-core JDBC key-value state

This PR introduces a new state backend (`DbStateBackend`) for storing partitioned and non-partitioned checkpoints in JDBC supported databases.

The main feature of this backend is the KvState implementation (`LazyDbKvState`) which provides consistent out-of-core storage of the key-value states for streaming tasks while only keeping a predefined number of states in memory as a cache. The KvState also supports database sharding and automatic compactions.

## Core changes

In order to implement all the necessary functionality some changes had to be applied to the current state interfaces:

 1. `StateBackend.initializeForCurrentJob(...)` now takes the Environment of the instantiating task as a parameter, which allows the backend to access more information about the task
 2. On recovery the `restoreState` methods of various runtime components now receive a global recovery timestamp (conceptually similar to the checkpoint timestamp)

(2) is used to bring the storage to a consistent state after restoring to avoid reading partially successful checkpoints, I will describe in detail later.

## DbStateBackend

### Non-partitioned states

This implementation of the StateBackend stores the serialized checkpoints in database tables as Blobs.
For every JobID a new database table is created to store the non-partitioned states:

* Table name: checkpoints_JobID
* Columns: checkpointId bigint, timestamp bigint, handleId bigint, checkpoint blob
* Primary key: handleId

A new handleId is generated for each non-partitioned state checkpoint which will be used to retrieve and clear up later. (this is trivial)

The DbStateBackend also supports plugging in arbitrary state backend implementation for storing the non-partitioned states, in which case only kv states will be checkpointed to the configured database.
`new DbStateBackend(config, new MemoryStateBackend())` -> will store non-partitioned states in memory. I think this would be the more common usage pattern as the db state storage is more interesting for the key-value states.

### Key-Value states

This LazyDbKvState implements kv-state backend that will only keep a limited amount of kv states in memory and stores the rest in the database.
For each key-value state in the job we create a new table to store the states. 

Table layout:

* Table name: StateName_OperatorId_JobId
* Columns:  timestamp bigint, k varbinary(256), v blob
* Primary key: (k, timestamp)

As the columns suggest the database will hold multiple versions of a key-value pair versioned by their timestamp.

If a key is not in the KvState cache, it is retrieved from the database by selecting v with the largest timestamp that is smaller than the last written timestamp.

Whenever a LazyDbKvState takes a snapshot, it saves all its modified state (the cache keeps track of what has been modified since the last snapshot) in the database with the checkpoint timestamp and updates its last written ts. The returned KvStateSnapshot basically only needs to contain the stateId(jobid+opid+name) and the checkpoint timestamp (and some additional information for better compaction) so this is stored in-memory.

When the cache reaches its size limit, a predefined number of least recently accessed states are checkpointed to the database (only if they were modified) with timestamp = lastTimestamp+1. Subsequent "spills" will overwrite each other as they have the same timestamp.

#### Restoring the LazyDbKvState from the snapshot

As I mentioned, the snapshot contains stateId and checkpoint timestamp. We will restore the LazyDbKvState from this information after cleaning up the database of the failed snapshots.

We remove the failed records based on their timestamp: we remove every state update that has ts larger than the checkpoint ts and smaller than the recovery ts. This can be done as long as the recovery timestamp is larger than the last checkpoint timestamp produced by the coordinator. If the checkpoint coordinator survives than this is trivially true, and is still a reasonable assumption in case of master failure.

#### State compaction

By default no state compaction happens. The user can enable automatic compaction at every so many successful checkpoints by calling dbConfig.setKvStateCompactionFrequency(int compactEvery). Compaction removes all k,v pairs for which there is a more recent valid checkpoint.

This should be made tunable if we want to continue a job from a previous checkpoint.

### DbAdapter

As almost all databases have slightly different SQL semantics it's very hard to write universal sql code that will execute on all of them in the same way. To bridge api differences the `DbAdapter` interface can be implemented which defines how each database operation is executed on the connection (how to insert, lookup or clean up states etc).

In this PR there 2 adapter implementations: MySqlAdapter, DerbyAdapter (for tests).

### Configuration

The DbStateBackend parameters can be configured by passing a `DbBackendConfiguration` object which will hold all necessary information about the database connection, number of shards, maximum cache size for the tasks, maximum batch insert size etc.

### Sharding

Database sharding is supported and the user can provide more than one database url to the DbBackendConfig. By default each key is written to the shard indexed by keyHash % numShards. It is also possible to define an custom partitioning strategy by passing a Partitioner instance in the DbBackendConfig.setPartitioner(...) method.

To handle sharding each task will maintain exactly 1 connection to each of the shards and keys are partitioned and inserted/ accordingly.

### Tests

The backend's functionality is tested using similar testing logic to the preexisting backend. Unit test for the backend and an IT case for running a checkpointed job. All the tests use Apache Derby as their underlying database.

I also ran some tests using MySql. I haven't found any obvious issues, I could store/access 100m+ kv states with about 256 mb tm heap. Insert/lookup query execution time seems to be reasonable with inserts taking under 100 microsecs and lookups in the order of 1-200 microsecs (this depends a lot on the job but I would assume it stays around  this with compaction).

We also ran some cluster experiments with about 20 database shards and couple hundred million keys.

## Possible improvements

### Use bloom filters to detect new keys

Currently new keys will be looked up in the database which adds an unnecessary round-trip to each new key seen, considerably slowing down processing of these elements. A way to tackle this is to keep a bloom filter at each KvState instance which would hold all keys that have been written to the database so far.

The limitation of this approach is that a bloom filter may be relatively big (couple hundred mbs) for a large number of keys (and it cannot be resized later), which might slow down the checkpointing from the beginning. Additionally the bloom filters probably need to be stored using the FsStateBackend.

aljoscha commented:
Wow, a lot of stuff. I will look into it once the release is out. :smiley:

gyfora commented:
My last commit (that was meant to solve the problems with failed tasks writing to the db) introduced some issues with the exactly once guarantees. I will look into it tomorrow.

StephanEwen commented:
Good stuff! Will need a day more to look through this, but this is a cool way of doing stateful stream computation :-)

uce commented:
Thanks for the great write up!

StephanEwen commented:
Cool stuff, really! This is very much in line with what I had in mind for a SQL backend.

Let me check if I understood everything correct (and see where my understanding is wrong), because I think we should be able to make an "exactly once" version of this based that mechanism. I am basically rephrasing what you describe in a different model.

### Basic Mode

What this is effectively doing is a batched and asynchronous version of distributed 2-phase commit transactions. The phases look basically like this:

  - **Adding data**: Pipe all modifications into the database, but not commit the transaction. They are tagged with the timestamp of the upcoming checkpoint (or any coordinated increasing version counter). This can happen in the background thread, for as long as the in-operator cache holds all edits that are not in the database yet.

  - **Pre-commit**: This is when the checkpoint is triggered. All pending edits are written into the database and then the transaction is committed. The state handle only includes the timestamp used on the elements. In the classical 2-phase transactions, after a task acks the pre-commit, it has to be able to recover to that state, which is given here. The checkpoint is not immediately valid for recovery though, which means that recovery has to have either a filter, or issue a query that deletes all records with timestamps larger than the version given during recovery. After the pre-commit, the timestamp is locally incremented and work can continue.

  - **Full commit**: This happens implicitly when the checkpoint coordinator marks the checkpoint as complete.

  - **Recovery**: The timestamp (or version counter) of the last successful checkpoint is restored, the deletion of records that were committed (but where the checkpoint did not succeed as a whole) happens, then records are lazily fetched. 

So far, this should give exactly once guarantees, or am I overlooking something?

### Compacting

Whenever the "checkpoint complete" notification comes (or every so many changes) you trigger a clean-up query in the background. Given that the SQL database has a not completely terrible query planner, this SQL statement would be okay efficient (single semi join).
```
DELETE FROM "table name" t1
WHERE EXISTS 
  (SELECT *
     FROM "table name" t2
    WHERE t2.handle_id = t1.handle_id
      AND t2.timestamp > t1.timestamp    //-- a newer version exists for the same handle
      AND t2.timestamp <= GLOBAL_VERSION //-- and the newer version is globally committed
  )
```
The good thing is that by virtue of having the incrementing global versions, we can set the isolation level for the query to "read uncommitted", which means that it will not lock anything and thus not compete with any other ongoing modifications.

aljoscha commented:
Just some remarks:
 - `DummyEnvironment` seems unnecessary, we already have `StreamMockEnvironment`. I think it could be reused.

- In the first version you had both the timestamp and checkpoint and recovery/key lookup took both into account. The recent version uses just the timestamp for lookup. Both introduce the new restore timestamp in the restore methods.

- The cleanup of failed checkpoints took into account the checkpoint and the recovery timestamp, but I think the recovery timestamp was always redundant since the condition in the SQL statement would always hold.

=> I think the timestamp is not needed. Can't everything be implemented by just using the (monotonically rising) checkpoint IDs?

Also, this is unrelated but maybe @StephanEwen or @gyfora know: Why to we have the `allOrNothingState` in `CheckpointCoordinator.restoreLatestCheckpointedState`?

gyfora commented:
@aljoscha 
 1. I was initially using the MockEnvironments but I added the DummyEnvironment for several reasons: I wanted control over the JobId and the number of subtasks for which I would have changed the MockEnvironment. Also I wanted to avoid having to clean up the memorymanager and other resources as I really don't need them
 2. I don't really understand what you mean here, the recovery timestamp is only used for cleanup on restore
 3. Imagine a scenario where 2 task are restoring . 1 restores quickly and starts writing new timestamps. If we call cleanup on the other task it will remove the new states if we don't bound by recovery timestamp. This can happen easily.

I don't know about the allOrNothingState :/

aljoscha commented:
2. Ah, I meant the lookupTimestamp. In an earlier version you used both the checkpointId and lookupTimestamp to perform key lookups.
3. I see, in this implementation of state the timestamp has basically assumed the role of the checkpointId and the checkpointId is (I think) completely ignored. Correct? Couldn't we then change the semantics of the checkpointId to work like the timestamps (they are somewhat logical, not physical timestamps anyways)?

gyfora commented:
You are right about the checkpointIDs being ignored. I don't see why we need to bother with changing the id semantics, the timestamps also serve as checkpoint Ids perfectly (in fact we could drop the checkpoint ID everywhere in Flink and just keep the timestamp).

The good thing about timestamps is that they are not incremental, meaning that if the first checkpoint have ts = 100 and the second has ts = 5100 then we can write the intermediate updates with ts 101, 102 ... while maintaining uniqueness and monotonicity. This is something that we use here and that is why I am using timestamps instead of ids.

gyfora commented:
@StephanEwen 
Thanks for the comments. You are right the main idea is exactly as you described. 
The reason why exactly-once is violated in some corner cases because it can happen that the pre-commit phase of the previous checkpoint is still failing during recovery. 

If we assume that the previous job is completely killed of, no writing to the database whatsoever after that happens, then we can properly clean up during recovery. 

This unfortunately does not seem to hold if you set the retry wait time to very low (like 0 ms in the snapshot). What this means is that the failed job is still writing the failed snapshot to the database after you recovered and cleaned up.

As for the compaction, I came up with something very similar for compaction but here is the funny thing and my problem. The query you wrote will run properly on Derby but is invalid on MySql (you cannot create a subquery for the same table as you are modifying). In mysql you need to create an inner join, but that will not work in Derby :P 

In any case I have made a prototype of this on: https://github.com/gyfora/flink/tree/compaction
The user can define the frequency of compaction (compact every so many checkpoints). And it also makes sure that compaction and cleanup is only executed on 1 subtask to avoid double work.

aljoscha commented:
@gyfora That's what I meant, basically the timestamps could subsume the role of the checkpointIds. I.e. The checkpointIds have the semantics of the timestamps and the timestamps would not be required. (Or the checkpointId would be removed and timestamps remain, depends on how you look at it... :smile: )

gyfora commented:
My last commit introduces automatic compaction with user specified frequency. It also allows the KvStates to implement the CheckpointNotifier interface in which case they will also get notified.

gyfora commented:
After some initial discussion we @StephanEwen we came to the following conclusions:

The current timestamp based approach has some limitations in terms of the assumptions it makes about the clocks on the checkpoint coordinator which can cause some issues in case of master failure. It is currently assumed that the recovery timestamp assigned by the checkpoint coordinator is larger than the last checkpoint timestamp assigned before failure. If the clock difference is large enough this might be a problem, although fairly unlikely.

The alternative approach would be to store the id, key, value triplets in the database (drop the timestamps). Snapshots written between two snapshots would be written with an incremented checkpoint id (and multiple spills to the same key would update the value not insert a new one every time). Cleanup could then be performed by getting the next checkpoint id from the coordinator (which would get it from zookeper). The drawback in this case would be an increased complexity for the batch inserts as we need to be able to handle primary key conflicts efficiently. While this is easy on some databases  (in MySql) in other cases (Derby) can be problematic.

gyfora commented:
I am now writing a prototype for this version, the batch insert got pretty ugly...
I will probably finish it tomorrow.

gyfora commented:
I updated this PR with the reworked logic, it has several advantages over the previous timestamp based solution (including the elimination of transactions from the logic).

The only problem I see is that the derby batch inserts happen row by row currently as it does not have the insert or update semantics.

This rework will also make it easier to write connectors to non transactional stores like Cassandra

gyfora commented:
I also removed the sharding logic now as I think it was pretty weak and not very useful (it maintained 1 connection per subtask which would break if we change parallelism)

gyfora commented:
I updated the sharding logic to do mod hashing by default on the keys for the number of shards, and the user can also add a custom Partitioner to implement custom sharding strategy.

gyfora commented:
Should we do a final iteration over this and merge this to contrib?

The description got slightly out of date when I changed this back so that it stores the state by timestamp (but its basically ctr+f replace id with timestamp)

gyfora commented:
I would like to push this soon if no objections

uce commented:
I agree that this will be an important backend and good to have in. :) But do we need to push this right now? I think we should wait a little and make sure that it fits well into the other ongoing changes I think Stephan and Aljoscha are working on. At least let's wait for someone to review it again.

gyfora commented:
Well, I don't know what they are working on... It would be easier not having to rebase state backend api changes

uce commented:
I totally understand your point, but I think it's OK that changes of this scope take longer to review and get in (my HA PR took over a month or so to get in). At the end of the day, it matters more that we get this right (because it covers a very important use case) than getting it in a few days earlier.

gyfora commented:
I have updated the description, and ran some more cluster tests without any issues.

It would be good if you all could do a second round of reviews please.

aljoscha commented:
I'm looking at it again.

StephanEwen commented:
Looking though this again...

StephanEwen commented on flink-contrib/flink-streaming-contrib/src/main/java/org/apache/flink/contrib/streaming/state/DbBackendConfig.java:
```diff
@@ -0,0 +1,406 @@
+	}
+
+	@Override
+	public boolean equals(Object obj) {
```
Overriding equals() but not hashCode(). Not sure what the usage of this object is, but dropping the equals method may make sense (do you ever compare configs fro equality?).

gyfora replied:
That must have been a mistake I generated it with eclipse :/ I only used it for a test but it doesnt make too much sense so I can remive it.

StephanEwen commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/StateBackend.java:
```diff
@@ -64,33 +64,35 @@
 	 * @throws Exception Exceptions may occur during initialization of the state and should be forwarded.
 	 */
-	public abstract <K, V> KvState<K, V, Backend> createKvState(
+	public abstract <K, V> KvState<K, V, Backend> createKvState(int operatorId, String stateName,
```
I would like to get rid of this change and simply let the state backend create a UID for the state name.

This method is called one per proper creation of a state (so it should not need deterministic state naming). Recovery happens from the state handle, which can store all required info.

gyfora replied:
I am not completely sure what you mean here.

Multiple different states can have the same name in different tasks. As far as I know we dont assume unique state names. This gets worse if the chained tasks have states with the same name then they actually go to the same backend as well.

I dont see how to go around this without an operator id. Could you please clarify your idea?

StephanEwen replied:
I suggest to not let the operator supply an ID and name, but simply leave the naming of the state to the state backend. The SqlStateBackend could just use `UUID.randomUUID().toString()` instead of `operatorId+stateName`.

gyfora replied:
The point is that all parallel instances write to the same set of tables. This will way sharding is transparently handled and the job parallelism can actually change without affecting the state. (No need to repartition it)

gyfora replied:
Otherwise you will have to create p*numShards tables and you wont even know what state is in it from looking at the table names

StephanEwen replied:
What you mention depends on the parallel subtask ID (which is already given in the initialize() method). The operatorId and name are the same for all parallel instances anyways.

StephanEwen replied:
The "name" (as a string) of the state is a very API specific thing that no other part of the runtime is concerned with. The operator ID is something specific to the StreamGraphBuilder and not to the streaming tasks at all. I think we are tying things together here that should not be tied together.

I still do not understand how this affects sharding. Does the shard assignment depend on the state name (rather than the parallel subtask / JobVertexId) ?

I only see that the table names will have the task name instead of the name of the state.

gyfora replied:
Let me first describe how sharding works than I will give a concrete example. 
Key-Value pairs are sharded by key not by the subtask. This means that each parallel subtask maintains a connection to all the shards and partitions the states before writing them to the appropriate shards according to the user defined partitioner (in the backend config). This is much better than sharding by subtask because we can later change the parallelism of the job without affecting the state and also lets us defined a more elaborate sharding strategy through the partitioner.

This means, when a kv state is created we create a table for that kvstate in each shard. If we would do it according to your suggestion we would need to create numShards number of tables for each parallel instance (total of p*ns) for each kvstate. Furthermore this makes the fancy sharding useless because we cannot change the job parallelism. So we need to make sure that parallel subtasks of a given operator write to the same state tables (so we only have ns number of tables regardless of the parallelism).

In order to do this we need something that uniqely identifies a given state in the streaming program (and parallel instances should have the same id).

The information required to create such unique state id is an identifier for the operator that has the state + the name of the state. (The information obtained from the environment is not enough because chained operators have the same environment, therefore if they have conflicting state names the id is not unique). The only thing that identifies an operator in the logical streaming program is the operator id assigned by the jobgraphbuilder (thats the whole point of having it). 

An example job with p=2 and numshards = 3:

chained map -> filter, both the mapper and filter has a state named "count", and let's assume that mapper has opid 1 and filter 2.

In this case the mapper would create 3 db tables (1 on each shard) with the same name kvstate_count_1_jobId. The filter would also create 3 tables with names: kvstate_count_2_jobId

All mapper instances would write to all three database shards, and the same goes for all the filters.

I hope you get what I am trying to say.

gyfora replied:
You might be right that state name and operator id is too api specific, but we will need ways to globally identify states which is impossible without that I think currently.

StephanEwen replied:
Thanks for the description of the sharding. The issue is that you need a deterministic table name that each KeyValueState can create independently.

StephanEwen commented:
I have a final comment inline. Otherwise, I think this is good to merge.

StephanEwen commented:
Had an offline chat with @gyfora with the following outcome:
  - A deterministic state identifier is needed
  - Small change to pass that identifier as a single ID String, initially internally constructed by state name + operator ID (as in this implementation)
  - That way, the streaming runtime can change handling of state names and operator IDs without breaking state backend implementations

With these changes, looks good to merge.

+1 from my side

rmetzger commented on flink-contrib/flink-streaming-contrib/src/main/java/org/apache/flink/contrib/streaming/state/DbStateHandle.java:
```diff
@@ -0,0 +1,88 @@
+			}, dbConfig.getMaxNumberOfSqlRetries(), dbConfig.getSleepBetweenSqlRetries());
+		} catch (IOException e) {
+			// We don't want to fail the job here, but log the error.
+			if (Log.isDebugEnabled()) {
```
I think you accidentally used Jetty's logging here (see `import org.eclipse.jetty.util.log.Log`)

gyfora replied:
Good catch, thanks Robert :)

uce replied:
In general: Can we add a check style import check for this? We have a Guava check to disallow importing from our shaded namespace already. @Robert: if you think that it is feasible, please open an issue for it.

rmetzger replied:
We could add a checkstyle rule for that, but I would like to solve the problem in a different way: I recently opened a JIRA for checking whether a Flink module is only using dependencies it has explicitly declared (forbidding to rely on transitive dependencies). WIth that check, we would also identify cases like this one.

gyfora commented:
@StephanEwen, @rmetzger:
I addressed the comments regarding the logs and the state id.

I also added a final improvement:

-Now compaction is executed in a background thread using a SingleThreadedExecutor
-At empty checkpoints a keepalive call is executed against the connections to avoid connection drops

These changes are in the last 2 commits, so if you guys +1 these last modifications I will merge it. I guess the compaction part is the most interesting here.

thanks

gyfora commented:
If no objections I would like to merge this :)

aljoscha commented:
I think you can go ahead. It's in contrib and you guys are battle-testing it anyways... :wink:

tillrohrmann commented on flink-contrib/flink-streaming-contrib/src/main/java/org/apache/flink/contrib/streaming/state/DbStateBackend.java:
```diff
@@ -0,0 +1,248 @@
+
+	// ------------------------------------------------------
+
+	private Environment env;
```
`env` is not serializable

tillrohrmann commented on flink-contrib/flink-streaming-contrib/src/main/java/org/apache/flink/contrib/streaming/state/DbStateBackend.java:
```diff
@@ -0,0 +1,248 @@
+	private final int numSqlRetries;
+	private final int sqlRetrySleep;
+
+	private PreparedStatement insertStatement;
```
`PreparedStatement` is not serializable.

tillrohrmann commented on flink-contrib/flink-streaming-contrib/src/main/java/org/apache/flink/contrib/streaming/state/DbStateBackend.java:
```diff
@@ -0,0 +1,248 @@
+ * {@link MySqlAdapter} can be supplied in the {@link DbBackendConfig}.
+ *
+ */
+public class DbStateBackend extends StateBackend<DbStateBackend> {
```
`StateBackend` implements the `Serializable` interface. Does this mean that `DbStateBackend` must also be `Serializable`? If this is the case, then this condition is violated because `env` and `insertStatement` are not serializable.
